File: stacks/go/net-http/rest/bff/flow/inference/blocked_on_review.go
package inference

import (
	"fmt"
	"idp-go-net-http-rest/bff/flow"
	"sort"
	"time"
)

func hoursElapsed(from string, now time.Time) float64 {
	parsed, err := time.Parse(time.RFC3339, from)
	if err != nil {
		return 0
	}
	return now.Sub(parsed).Hours()
}

func selectReviewState(states []flow.NormalizedReviewState) (flow.NormalizedReviewState, bool) {
	candidates := make([]flow.NormalizedReviewState, 0)
	for _, state := range states {
		if state.State == "awaiting_review" || state.State == "under_review" {
			candidates = append(candidates, state)
		}
	}
	if len(candidates) == 0 {
		return flow.NormalizedReviewState{}, false
	}
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].AsOf > candidates[j].AsOf
	})
	return candidates[0], true
}

func hasBlockingValidation(input flow.ProviderAdapterInput, changeID string) bool {
	for _, run := range input.ValidationRuns {
		if run.ChangeID == changeID && run.Scope == "branch" && (run.State == "failed" || run.State == "flaky") {
			return true
		}
	}
	return false
}

func resolveActors(actors []flow.NormalizedActor, ids []string) []string {
	if len(ids) == 0 {
		return nil
	}
	lookup := make(map[string]string, len(actors))
	for _, actor := range actors {
		lookup[actor.ProviderID] = actor.DisplayName
	}
	names := make([]string, 0, len(ids))
	for _, id := range ids {
		if name, ok := lookup[id]; ok && name != "" {
			names = append(names, name)
		}
	}
	sort.Strings(names)
	return names
}

func InferBlockedOnReview(input flow.ProviderAdapterInput, now time.Time) (*flow.FlowSignal, bool) {
	reviewState, ok := selectReviewState(input.ReviewStates)
	if !ok {
		return nil, false
	}

	refTime := reviewState.LastActivityAt
	if refTime == "" {
		refTime = reviewState.AsOf
	}

	elapsed := hoursElapsed(refTime, now)
	if elapsed < ReviewWindowHours {
		return nil, false
	}

	changeID := reviewState.ChangeID
	if hasBlockingValidation(input, changeID) {
		return nil, false
	}

	reviewers := resolveActors(input.Actors, reviewState.ReviewerActorIDs)
	waitingOn := "assigned reviewers"
	if len(reviewers) > 0 {
		waitingOn = fmt.Sprintf("%s", reviewers)
	} else if len(reviewState.ReviewerTeamNames) > 0 {
		waitingOn = fmt.Sprintf("%s", reviewState.ReviewerTeamNames)
	}

	partial := reviewState.IsPartial || input.Repository.IsPartial
	for _, c := range input.Changes {
		if c.ProviderID == changeID && c.IsPartial {
			partial = true
			break
		}
	}

	signal := flow.FlowSignal{
		ID:         "blocked_on_review",
		Title:      "Blocked on review",
		Severity:   "high",
		Confidence: degradeConfidence("high", partial),
		Explanation: fmt.Sprintf("Waiting %.0fh for review: pending %s.", elapsed, waitingOn),
		RecommendedNextAction: "Notify assigned reviewers or reassess reviewer assignment.",
		RelatedEntities:      []any{changeID},
	}

	return &signal, true
}

File: stacks/nodejs/react-fastify/rest/bff/src/flow/inference/blockedOnReview.ts
import { REVIEW_WINDOW_HOURS } from "./thresholds";
import { degradeConfidence, resolveActors, resolvePrimaryTeam } from "./explain";
import type {
  NormalizedActor,
  NormalizedReviewState,
  ProviderAdapterInput,
} from "../types";
import type { FlowSignal } from "./types";

function hoursElapsed(from: string, now: Date): number {
  const start = Date.parse(from);
  if (!Number.isFinite(start)) {
    return 0;
  }
  return (now.getTime() - start) / (1000 * 60 * 60);
}

function selectReviewState(reviewStates: NormalizedReviewState[]): NormalizedReviewState | null {
  const candidates = reviewStates.filter((state) =>
    state.state === "awaiting_review" || state.state === "under_review"
  );
  if (candidates.length === 0) {
    return null;
  }
  // Use most recent by as_of
  return candidates.sort((a, b) => Date.parse(b.as_of) - Date.parse(a.as_of))[0];
}

function hasBlockingValidation(input: ProviderAdapterInput, changeId: string): boolean {
  return input.validation_runs.some(
    (run) =>
      run.change_id === changeId &&
      run.scope === "branch" &&
      (run.state === "failed" || run.state === "flaky"),
  );
}

export function inferBlockedOnReview(
  input: ProviderAdapterInput,
  now: Date,
): FlowSignal | null {
  const reviewState = selectReviewState(input.review_states);
  if (!reviewState) {
    return null;
  }

  const referenceTime = reviewState.last_activity_at ?? reviewState.as_of;
  const elapsedHours = hoursElapsed(referenceTime, now);
  if (elapsedHours < REVIEW_WINDOW_HOURS) {
    return null;
  }

  const changeId = reviewState.change_id;
  if (hasBlockingValidation(input, changeId)) {
    return null;
  }

  const reviewers = resolveActors(input.actors, reviewState.reviewer_actor_ids);
  const reviewerTeams = reviewState.reviewer_team_names ?? [];
  const waitingOn =
    reviewers.length > 0
      ? reviewers.join(", ")
      : reviewerTeams.length > 0
        ? reviewerTeams.join(", ")
        : "assigned reviewers";

  const baseConfidence: FlowSignal["confidence"] = "high";
  const partialFlag =
    reviewState.is_partial === true ||
    input.repository.is_partial === true ||
    input.changes.some((c) => c.provider_id === changeId && c.is_partial);

  const confidence = degradeConfidence(baseConfidence, partialFlag);
  const service = input.repository.full_name ?? input.repository.provider_id;
  const team = resolvePrimaryTeam(input.ownership_hints);
  const observedAt = reviewState.as_of ?? referenceTime ?? new Date().toISOString();

  return {
    id: "blocked_on_review",
    title: "Blocked on review",
    severity: "high",
    confidence,
    explanation: `Waiting ${Math.floor(elapsedHours)}h for review: pending ${waitingOn}.`,
    recommendedNextAction: "Notify assigned reviewers or reassess reviewer assignment.",
    relatedEntities: [service, changeId],
    scope: { service, team, stage: "review" },
    observedAt,
  };
}
